compiler/resolution/visibleFunctions.cpp

```diff
@@ -147,6 +181,155 @@ static void buildVisibleFunctionMap() {
+                 isShadowVarSymbol(block->parentSymbol));
+      fnBlock = true;
+      BlockStmt* inFnInstantiationPoint = inFn->instantiationPoint();
+      if (inFnInstantiationPoint && !inFnInstantiationPoint->parentSymbol) {
+        INT_FATAL(inFn, "instantiation point not in tree\n"
+                        "try --break-on-remove-id %i and consider making\n"
+                        "that block scopeless",
+                        inFnInstantiationPoint->id);
+      }
+      if (inFnInstantiationPoint && inFnInstantiationPoint->parentSymbol)
+        instantiationPt = inFnInstantiationPoint;
+    }
+
+    if (call->id == breakOnResolveID) {
+      if (moduleBlock)
+        printf("visible methods: block %i  module %s  %s\n",
+               block->id, inMod->name, debugLoc(block));
+      else if (fnBlock)
+        printf("visible methods: block %i  fn %s  %s\n",
+               block->id, inFn->name, debugLoc(block));
+      else
+        printf("visible methods: block %i  %s\n",
+               block->id, debugLoc(block));
+
+      if (instantiationPt) {
+        printf("  instantiated from block %i  %s\n",
+               instantiationPt->id, debugLoc(instantiationPt));
+      }
+    }
+
+    // The following statement causes this to apply to all blocks,
+    // and not just module or function blocks.
+    //
+    // This is because e.g. in associative.chpl primer, instantiation occurs in
+    // a block that isn't a fn or module block.
+    visited.insert(block);
+
+    if (VisibleFunctionBlock* vfb = visibleFunctionMap.get(block)) {
+      // the block defines functions
+
+      if (Vec<FnSymbol*>* fns = vfb->visibleFunctions.get(name)) {
+        forv_Vec(FnSymbol, fn, *fns) {
+          // When private methods and fields are supported, we'll need to extend
+          // this
+          visibleFns.add(fn);
+        }
+      }
+    }
+
+    if (block->useList != NULL) {
+      // the block uses other modules
+      for_actuals(expr, block->useList) {
+        SymExpr* se = NULL;
+        if (UseStmt* use = toUseStmt(expr)) {
+          se = toSymExpr(use->src);
+        } else if (ImportStmt* import = toImportStmt(expr)) {
+          se = toSymExpr(import->src);
+        } else {
+          INT_FATAL("bad expr in useList, expected ImportStmt or UseStmt");
+        }
+        // Intentionally ignore use/import privacy and limitations
+        // Methods, fields, and special type support functions cannot be
+        // re-enabled via only lists (or doing so is not easy for the user).
+
+        // A use statement could be of an enum instead of a module, but only
+        // modules can define functions.
+        if (ModuleSymbol* mod = toModuleSymbol(se->symbol())) {
+          bool isVisible;
+          // Get potential scope pair.
+          std::pair<BlockStmt*, BlockStmt*> curPair =
+            std::make_pair(getVisibilityScope(call),
+                           mod->block);
+          // See if it's already in the map
```

Comment:
I don't understand why this map is necessary with this approach in the code. Aren't we just recursively visiting only those modules that are visible (by use statements etc)?

Reply:
The use statement could be of a private module, which would only be visible in the scope in which it was defined.

Reply:
In that case - couldn't you look at the ModuleSym and note that it is private and only then do the extra logic to sort out whether or not the call is within the module defining?

Reply:
Yeah, that sounds like a good idea!